### PR TITLE
[salesforce] Add `salesforce` package foundation along with `logout_rest` data stream

### PR DESCRIPTION
**Pull request by yug-elastic**
- Enhancement

## What does this PR do?

- Generated the skeleton of Salesforce integration package.
- Added 1 data stream ( `logout_rest` )
- Added data collection logic.
- Added the ingest pipelines.
- Mapped fields according to the ECS schema and added Fields metadata in the appropriate yml files.
- Added system test cases.

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/master/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] If I'm introducing a new feature, I have modified the Kibana version constraint in my package's `manifest.yml` file to point to the latest Elastic stack release (e.g. `^7.13.0`).

## How to test this PR locally

- Clone integrations repo.
- Install elastic-package locally.
- Start elastic stack using elastic-package.
- Move to integrations/packages/salesforce directory.
- Run the following command to run tests. 
> elastic-package test 

image

### CODE REVIEW

**Review by kush-elastic — APPROVED** [review]
LGTM

File: packages/salesforce/data_stream/logout_rest/fields/fields.yml [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+              description: The type of API request.
+            - name: version
+              type: keyword
+              description: The version of the API that's being used.
+        - name: app_type
+          type: keyword
+          description: The application type that was in use upon logging out.
+        - name: browser_type
+          type: keyword
+          description: The identifier string returned by the browser used at login.
```
Comment (agithomas) [inline packages/salesforce/data_stream/logout_rest/fields/fields.yml:27]: Change from login -> logout

File: packages/salesforce/data_stream/logout_rest/fields/fields.yml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+          description: Mode of API from which the event is collected.
+        - name: api
+          type: group
+          fields:
+            - name: type
+              type: keyword
+              description: The type of API request.
+            - name: version
+              type: keyword
+              description: The version of the API that's being used.
```
Comment (agithomas) [inline packages/salesforce/data_stream/logout_rest/fields/fields.yml:21]: Please change as Salesforce API

File: packages/salesforce/data_stream/logout_rest/fields/fields.yml [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+          description: The identifier string returned by the browser used at login.
+        - name: client_version
+          type: keyword
+          description: The version of the client that was in use upon logging out.
+        - name: event_type
+          type: keyword
+          description: The type of event. The value is always Logout.
+        - name: login_key
+          type: keyword
+          description: The string that ties together all events in a given user's login session. It starts with a login event and ends with either a logout event or the user session expiring.
```
Comment (agithomas) [inline packages/salesforce/data_stream/logout_rest/fields/fields.yml:36]: Change from login -> logout

File: packages/salesforce/data_stream/logout_rest/fields/fields.yml [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+          description: The code for the client platform. If a timeout caused the logout, this field is null.
+        - name: resolution_type
+          type: keyword
+          description: TThe screen resolution of the client. If a timeout caused the logout, this field is null.
+        - name: session
+          type: group
+          fields:
+            - name: level
+              type: keyword
+              description: The security level of the session that was used when logging out.
```
Comment (agithomas) [inline packages/salesforce/data_stream/logout_rest/fields/fields.yml:51]: Please consider adding the various levels as example (if it is a limited list)

File: packages/salesforce/data_stream/logout_rest/fields/fields.yml [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+            - name: type
+              type: keyword
+              description: The session type that was used when logging out.
```
Comment (agithomas) [inline packages/salesforce/data_stream/logout_rest/fields/fields.yml:54]: Possible to include the example values ?

**Review by agithomas — COMMENTED** [review]
Please also consider changing the title of the panel as

Top 10 IP address by Logout request count

File: packages/salesforce/_dev/deploy/docker/files/config.yml [hunk 6]
```diff
@@ -68,4 +68,4 @@ rules:
           content-type: ["text/csv"]
         body: |-
           "EVENT_TYPE","TIMESTAMP","REQUEST_ID","ORGANIZATION_ID","USER_ID","RUN_TIME","CPU_TIME","URI","SESSION_KEY","LOGIN_KEY","TYPE","METHOD","SUCCESS","TIME","REQUEST_SIZE","RESPONSE_SIZE","URL","TIMESTAMP_DERIVED","USER_ID_DERIVED","CLIENT_IP","URI_ID_DERIVED"
-          "ApexCallout","20221122044615.591","ABCDE","00D5j000000VABC","0055j000000ABCD","1305","10","CALLOUT-LOG","ABCDEF","ABCDEFGH","OData","GET","1","1293","10","256","https://temp.sh/odata/Accounts","2022-11-22T04:46:15.591Z","0055j012345utlPAAQ","127.0.0.1","0055j000000utlPABCD"
+          "ApexCallout","20221122044615.591","ABCDE","00D5j000000VABC","0055j000000ABCD","1305","10","CALLOUT-LOG","ABCDEF","ABCDEFGH","OData","GET","1","1293","10","256","https://temp.sh/odata/Accounts","2022-11-22T04:46:15.591Z","0055j012345utlPAAQ","81.2.69.142","0055j000000utlPABCD"
```
Comment (agithomas) [inline packages/salesforce/_dev/deploy/docker/files/config.yml:71]: I hope that the values here are copied from the actual logs received from an integration. And, i see ABCDEF, ABCDEFGH ,ABCDE etc in the sample. It is important to keep near realistic values in the mock data . Please use actual logs received here. You may consider masking any sensitive information if needed  (eg: actual token key , password, email etc)

**Comment by kush-elastic** [comment]
> 



> by Logout request count



> Please also consider changing the title of the panel as
> 
> Top 10 IP address by Logout request count

-> We have currently `Top 10 IP addresses by request count` as the entire dashboard shows the Logout data, Do you still suggest to get this updated ? The only idea of not adding Logout in the name is that the dashboard shows only Logout data.

**Comment by kush-elastic** [comment]
@agithomas We have followed this official documentation from [Salesforce](https://developer.salesforce.com/docs/atlas.en-us.object_reference.meta/object_reference/sforce_api_objects_eventlogfile_logout.htm) for all the fields descriptions. We have attached same link in README.md Please have a look over this once.

**Comment by agithomas** [comment]
> @agithomas We have followed this official documentation from [Salesforce](https://developer.salesforce.com/docs/atlas.en-us.object_reference.meta/object_reference/sforce_api_objects_eventlogfile_logout.htm) for all the fields descriptions. We have attached same link in README.md Please have a look over this once.

Some of the suggestions are correct from Elastic's documentation perspective. If you find any suggestions conflicting, please point out.

File: packages/salesforce/data_stream/logout_rest/fields/fields.yml (hunk 1 above)
Comment (kush-elastic) [reply packages/salesforce/data_stream/logout_rest/fields/fields.yml:27]: I would suggest to keep it to `login` as per the salesforce [documentation](https://developer.salesforce.com/docs/atlas.en-us.object_reference.meta/object_reference/sforce_api_objects_eventlogfile_logout.htm#:~:text=The%20identifier%20string%20returned%20by%20the%20browser%20used%20at%20login.) for the field `browser_type`.
Let me know if you still suggest to update it

File: packages/salesforce/data_stream/logout_rest/fields/fields.yml (hunk 5 above)
Comment (kush-elastic) [reply packages/salesforce/data_stream/logout_rest/fields/fields.yml:54]: As the example [value](https://developer.salesforce.com/docs/atlas.en-us.object_reference.meta/object_reference/sforce_api_objects_eventlogfile_logout.htm#:~:text=High%2DAssurance%20Session-,SESSION_TYPE,-Type) list is long will update it with few of them
`The session type that was used when logging out (e.g. API, Oauth2 or UI).`

Let me know if you insist to add all the example values.

**Review by SubhrataK — APPROVED** [review]

**Comment by agithomas** [comment]
> We have currently `Top 10 IP addresses by request count` as the entire dashboard shows the Logout data, Do you still suggest to get this updated ? The only idea of not adding Logout in the name is that the dashboard shows only Logout data.

Yes, please

File: packages/salesforce/data_stream/logout_rest/elasticsearch/ingest_pipeline/default.yml [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+    fields:
+    - json.REQUEST_ID
+    - json.SESSION_KEY
+    - json.TIMESTAMP_DERIVED
+    - json.USER_ID
+    target_field: _id
+    ignore_missing: true
+- set:
+    field: salesforce.logout.access_mode
+    value: "rest"
```
Comment (agithomas) [inline packages/salesforce/data_stream/logout_rest/elasticsearch/ingest_pipeline/default.yml:23]: Recommend to change this value to REST . If possible consider changing it in login_rest datastream as well.

File: packages/salesforce/data_stream/logout_rest/elasticsearch/ingest_pipeline/default.yml [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+    ignore_failure: true
+    ignore_empty_value: true
+- set:
+    field: ecs.version
+    value: "8.4.0"
```
Comment (agithomas) [inline packages/salesforce/data_stream/logout_rest/elasticsearch/ingest_pipeline/default.yml:28]: Consider using the latest ecs.version

**Review by agithomas — COMMENTED** [review]
Recommend to include a Geo Map similar to login_rest datastream to indicate the logout geo locations making use of the geo lat & lon fields

File: packages/salesforce/data_stream/logout_rest/elasticsearch/ingest_pipeline/default.yml (hunk 8 above)
Comment (kush-elastic) [reply packages/salesforce/data_stream/logout_rest/elasticsearch/ingest_pipeline/default.yml:28]: When we started creating the package we had `8.4.0` as the latest release,
Should we update it to 8.5.0(Current stable release) or 8.6.0(Future release). Please let us know.

Reference for 8.5.0: https://www.elastic.co/guide/en/ecs/8.5/index.html
Reference for 8.6.0: https://www.elastic.co/guide/en/ecs/8.6/index.html

Comment (agithomas) [reply packages/salesforce/data_stream/logout_rest/elasticsearch/ingest_pipeline/default.yml:28]: 8.5 is good enough

**Review by agithomas — APPROVED** [review]
LGTM

**Pull request merged**